### PR TITLE
VF2 Layout : The layout allocation as a subgraph isomorphism problem

### PR DESCRIPTION
**Pull request by 1ucian0**
~On hold, depends on https://github.com/Qiskit/retworkx/pull/368 and https://github.com/Qiskit/retworkx/pull/375~

Exploring alternatives with @czachow on how to improve `CSPLayout`, we start reading more about subgraph isomorphism problem, in particular VF2. In this way, we noticed that, when VF2 finds an isomorphism a subgraph of G and H, also finds the mapping _m_ that pairs the nodes in H with nodes in G. That mapping *is* they layout. @mtreinish and @georgios-ts [wrote a `is_subgraph_isomorphic` for retworkx some time ago](https://github.com/Qiskit/retworkx/pull/317) using VF2 and they *just* [need to expose *m*](https://github.com/Qiskit/retworkx/pull/368). This pass uses *m* to construct a matching subgraph, like `CSPLayout` but much faster.

## Features

This pass has the following features:
 * Handles direction: because retworkx supports `digraph_vf2_mapping` and `graph_vf2_mapping`.
 * Introduces non-determinism by shuffling the coupling map nodes labels. In this way, if there are many isomorphic subgraphs, will get a random one. That can be deactivated with `seed=-1`.
 * ~The parameter `id_order` is passed to the underling `*graph_vf2_mapping` call. If `None` activates id_order only if the problem is small (a subgraph to find with less than 300 edges).~
 *  It works very much like `CSPLayout`, but faster.

## Performance 

In order to illustrate the performance advantage, I used `timeit` in the following example:
```python
from qiskit import QuantumRegister, QuantumCircuit
from qiskit.transpiler import CouplingMap
from qiskit.transpiler.passes import VF2Layout, CSPLayout
from qiskit.converters import circuit_to_dag
from qiskit.test.mock import FakeTenerife
from timeit import timeit

def run_pass(qc, pass_):
    dag = circuit_to_dag(qc)
    return timeit(lambda: pass_.run(dag), number=100)

tenerife_cm = FakeTenerife().configuration().coupling_map

qr = QuantumRegister(3, "qr")
circuit = QuantumCircuit(qr)
circuit.cx(qr[1], qr[0])  # qr1 -> qr0
circuit.cx(qr[0], qr[2])  # qr0 -> qr2
circuit.cx(qr[1], qr[2])  # qr1 -> qr2

print('Tenerife:')
print('VF2Layout:', run_pass(circuit, VF2Layout(CouplingMap(tenerife_cm))))
print('CSPLayout:', run_pass(circuit, CSPLayout(CouplingMap(tenerife_cm), time_limit=None, call_limit=None)))
```
The result for this small case is noticible:
```
Tenerife:
VF2Layout: 0.00985708399999985
CSPLayout: 0.015390594999999951
```
With the example from https://github.com/Qiskit/qiskit-terra/issues/5694 (where a trivial layout in Manhattan makes CSP timeout in level 2 and 3), the difference is huge:
```python
import numpy
from qiskit.circuit.library import GraphState
from qiskit.test.mock import FakeManhattan


backend = FakeManhattan()
config = backend.configuration()
manhattan_cm = FakeManhattan().configuration().coupling_map

rows = [x[0] for x in manhattan_cm]
cols = [x[1] for x in manhattan_cm]

A = numpy.zeros((65, 65))
A[rows, cols] = 1

circuit = GraphState(A).decompose()

print('Manhattan:')
print('VF2Layout:', run_pass(circuit, VF2Layout(CouplingMap(manhattan_cm))))
print('CSPLayout:', run_pass(circuit, CSPLayout(CouplingMap(manhattan_cm), time_limit=None, call_limit=None)))
```
```
Manhattan:
VF2Layout: 0.16880208799999963
CSPLayout: 60499.96817250001
```

### Noise awareness
 Since this is so fast, probably would be better to add noise awareness a la https://github.com/Qiskit/qiskit-terra/pull/5075. For that, several possible solutions can be found and then calculate the fidelity. This can be done by a fidelity scoring pass (to add later)

### CODE REVIEW

**Comment by mtreinish** [comment]
If you update the retworkx line in the requirements.txt file to:

```
retworkx @ git+https://github.com/mtreinish/retworkx@vf2-mapping
```

Then you can run the tests in CI. See: https://github.com/Qiskit/qiskit-terra/pull/6302/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552 for where I'm doing that on another PR that depends on an in progress retworkx commit.

File: qiskit/transpiler/passes/layout/vf2_layout.py
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/vf2_layout.py:54]: I would avoid doing it this way as using index in a loop will potentially traverse the list fully every time. Instead I would do a single iteration then just do a dict lookup to get the index with something like:

```suggestion
        qubit_indices = {qubit: index for index, qubit in enumerate(qubits)}
        interactions = [(qubit_indices[gate.qargs[0]], qubit_indices[gate.qargs[1]]) for gate in dag.two_qubit_ops()]
```

```python

```

Comment (1ucian0) [reply qiskit/transpiler/passes/layout/vf2_layout.py:54]: It makes sense, thanks! (this code is also in `CSPLayout`).

**Comment by georgios-ts** [comment]
Hi @1ucian0,  i'm afraid we need to tweak a bit `is_subgraph_isomorphic` function to find layouts since it currently works for the *induced* subgraph case. What this means is the following will return `False`:
```python
import retworkx as rx

ga = rx.PyGraph()
ga.add_nodes_from([0, 1, 2])
ga.add_edges_from_no_data([
    (0, 1), (1, 2), (2, 0)
])

gb = rx.PyGraph()
gb.add_nodes_from([0, 1, 2])
gb.add_edges_from_no_data([
    (0, 1), (1, 2)
])

rx.is_subgraph_isomorphic(ga, gb)
---
False
```
because the edge `(0, 2)` does not exist in the second graph.
Sorry for the confusion, i should have document this better in the first place. Good news, it shouldn't be a big change and i will try to push a PR asap

File: releasenotes/notes/vf2layout-4cea88087c355769.yaml
Comment (mtreinish) [inline releasenotes/notes/vf2layout-4cea88087c355769.yaml:6]: This is only true if you set `id_order=False`. The vf2++ heuristic matching order is not used when you call `vf2_mapping()` with `id_order=True`.

Comment (1ucian0) [reply releasenotes/notes/vf2layout-4cea88087c355769.yaml:6]:
```suggestion
    isomorphism problem. For this, it uses VF2, a state of the art heuristic to find
```

**Comment by 1ucian0** [comment]
> Hi @1ucian0, i'm afraid we need to tweak a bit `is_subgraph_isomorphic` function to find layouts since it currently works for the _induced_ subgraph case.

Adding a layout version of the same situation as a test... Once https://github.com/Qiskit/retworkx/pull/372 is merged, I can add the related adjustment here. Thanks!

File: qiskit/transpiler/passes/layout/vf2_layout.py
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/vf2_layout.py:81]:
```suggestion
        mapping = vf2_mapping(cm_graph, im_graph, subgraph=True, id_order=True, induced=False)
```

**Comment by 1ucian0** [comment]
When `id_order=False`, even with node randomization, the algorithm always find the same mapping when the problem is "small". With "big" problems, `id_order=True` [takes too long](https://github.com/Qiskit/retworkx/pull/368#issuecomment-869131142).

In this version of the pass, I dynamically deactivate `id_order` when the subgraph to find has more than 300 edges.

Does it make sense @georgios-ts @mtreinish ?

**Comment by georgios-ts** [comment]
@1ucian0 Yes, this is an expected behavior. With `id_order=false` we reorder the nodes based on a heuristic that depends first on the graph topology and secondly on node ids. I guess in "small" graphs you tried, we don't have "draws" to break based on node ids so randomization doesn't have any effect.

**Comment by 1ucian0** [comment]
So my solution seems reasonable?

**Comment by 1ucian0** [comment]
removing `on hold` as retworkx is ready

**Review by mtreinish — DISMISSED** [review]
This LGTM now, thanks for the updates. Before merging though I have 1 question inline about the docs and some nits in the tests (nothing blocking there though).

File: test/python/transpiler/test_vf2_layout.py
Comment (1ucian0) [reply test/python/transpiler/test_vf2_layout.py]: merged in 20f044d01f5583f40fedb772af157ae143d30841

**Comment by 1ucian0** [comment]
> Per #7155 I'm wondering if we want to mention here that this only takes into account 2q gates.

oh gosh. Indeed. Actually, this unveiled that most of the tests were doing nothing, as they graph gate is >2q. Fixed in b7eb52ceaa85e631204c19680894a91974bfaea5 . 

> Nothing layout/routing related works for >2q right now, but when #5885 merges the transpiler can potentially know that it has a 3q basis gate and this pass won't be compatible with that ever (can't represent a >2q link as a graph).

I guess it's a problem for later. For now, the pass only works with 2q. Anything >2q, raises.

File: qiskit/transpiler/passes/layout/vf2_layout.py
Comment (1ucian0) [reply qiskit/transpiler/passes/layout/vf2_layout.py:26]: see https://github.com/Qiskit/qiskit-terra/pull/6620#issuecomment-957226964

File: test/python/transpiler/test_vf2_layout.py
Comment (1ucian0) [reply test/python/transpiler/test_vf2_layout.py:283]: I think is better to mirror the issue code as much as possible.

File: qiskit/transpiler/passes/layout/vf2_layout.py
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/vf2_layout.py:44]:
```suggestion
"""VF2Layout pass to find a layout using subgraph isomorphism"""
import random
from retworkx import PyGraph, PyDiGraph, vf2_mapping
from qiskit.transpiler.layout import Layout
from qiskit.transpiler.basepasses import AnalysisPass
from qiskit.transpiler.exceptions import TranspilerError


class VF2Layout(AnalysisPass):
    """A pass for choosing a Layout of a circuit onto a Coupling graph, as a
    a subgraph isomorphism problem, solved by VF2++.
        
    If a solution is found that means there is a "perfect layout" and that no
    further swap mapping or routing is needed. If a solution is found the layout
    will be set in the property set as ``property_set['layout']``. However, if no
    solution is found, no ``property_set['layout']`` is set. The stopping reason is
    set in ``property_set['VF2Layout_stop_reason']`` in all the cases and will be
    one of the following values:

        * ``"solution found"``: If a perfect layout was found.
        * ``"nonexistent solution"``: If no perfect layout was found.

    """

    def __init__(self, coupling_map, strict_direction=False, seed=None):
        """Initialize a ``VF2Layout`` pass instance

        Args:
            coupling_map (CouplingMap): Directed graph representing a coupling map.
            strict_direction (bool): If True, considers the direction of the coupling map.
                                     Default is False.
            seed (int): Sets the seed of the PRNG. -1 Means no node shuffling.
        """
```

**Review by mtreinish — COMMENTED** [review]
I just realized the docstrings need to be updated here, because the rendered docs would contain no detail on how the pass worked. I left a suggestion inline on how to fix it (hopefuilly I got the spacing right in the github editor). After that's fixed I think this is ready.

**Review by mtreinish — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
